### ISSUE TITLE
[Bug][pypto-lib] mtp/hc_head reads a full 16-row tile from an 8-row tensor and segfaults on both simulators

### ISSUE THREAD

**Issue by Crystal-wzy** [hw-native-sys/pypto-lib#1111]
### Diagnosis

`**pypto-lib / models/deepseek_v4_flash_mtp/hc_head.py`** — The `hc_head_linear` stage rounds the row count up to the Cube fractal M-tile and then slices that many rows without bounding the read to the rows that actually exist.

```python
# models/deepseek_v4_flash_mtp/hc_head.py:53
t_linear = pl.max(t_dim, LINEAR_T_TILE)          # 8 -> 16
...
# models/deepseek_v4_flash_mtp/hc_head.py:89
x_lin = x_flat[t0 : t0 + LINEAR_T_TILE, k0 : k0 + LINEAR_K_TILE]   # no valid_shape
w = pl.slice(hc_head_fn, [HC_PAD, LINEAR_K_TILE], [0, k0],
             valid_shape=[HC_MULT, LINEAR_K_TILE])                 # the weight slice has one
```

For the MTP decode configuration `t_dim = DECODE_BATCH * DECODE_SEQ = 4 * 2 = 8`, while `LINEAR_T_TILE = 16`. The generated Cube kernel therefore emits a `GlobalTensor` claiming 16 rows over an 8-row tensor and loads rows 8..15, which do not exist.

This is a genuine out-of-bounds read, not a simulator artifact. It has two distinct manifestations:

1. **Crash** when the token count is below the tile (`t_dim < LINEAR_T_TILE`): the load runs off the end of the allocation.
2. **Silent wrong results** when the token count is above the tile but not a multiple of it: `pl.max` does not round up to a multiple, so `t_linear // LINEAR_T_TILE` drops the trailing rows. The seed and linear tasks never cover them, `mixes_raw` keeps uninitialised values there, and the reduce stage reads them.

The sibling DSpark implementation is already correct and is the model for the fix: it separates aligned blocks from a single trailing partial block and gives the tail an explicit `valid_shape` (`models/deepseek_v4_flash_dspark/hc_head.py:64-65, 106-165`).

### Source

- Parent CI run: https://github.com/hw-native-sys/pypto-lib/actions/runs/33453529066
- Workflow: `Daily CI #187` (scheduled 2026-09-01 00:08 UTC)
- pypto-lib branch / commit: `main` / `0e7e07aa5e42db6e08c68fd5bfe1def571492b53`
- Jobs: `model-tests-sim (a2a3sim)`, `model-tests-sim (a5sim)`
- Affected case: `models/deepseek_v4_flash_mtp/hc_head.py`
- Contrast case: `models/deepseek_v4_flash_dspark/hc_head.py` — passes on `a2a3sim` in the same run
- Real `a2a3`: the case is **not** in the device-job failure list, so hardware passes

### Failure Details

Crash stack, captured by rebuilding the JIT kernels with `-g -O0` and running under `gdb`:

```text
#0 pto::TLOAD_TILE_IMPL<Tile<Mat, float, 16, 256, ColMajor, -1, -1, RowMajor, 512, ...>,
                        GlobalTensor<float, Shape<1,1,1,16,256>,
                                     Stride<262144,262144,262144,16384,1>, Layout::ND>>
#1 pto::TLOAD_IMPL<...>
#2 pto::TLOAD<...>
#3 hc_head_linear (build_output/.../kernels/aic/hc_head_linear.cpp)

si_addr = 0xaaaab03cfee0
[heap] ends at 0xaaaab03cc000  ->  fault is 16096 bytes past the end of the heap
```

Arithmetic of the overread:

| Quantity | Value |
|---|---:|
| `HC_DIM = hc_mult * hidden_size` | 16384 |
| `x_flat` shape / size | `[8, 16384]` FP32, 524288 bytes |
| elements in `x_flat` | 131072 |
| highest element the kernel reads (`15 * 16384 + 255`) | 246015 |
| overread | 114944 elements = 459776 bytes ≈ 449 KB |

Why hardware survives it: on A2/A3 the overread lands in other mapped global memory, and the garbage rows are written into `mixes_raw` rows 8..15, which the reduce stage never reads back (it iterates `t_dim // T_TILE` tiles, rows 0..7). Under CPU_SIM the tensor is a host allocation of exactly 8 rows, so the same read touches an unmapped page.

### Minimal Reproduction

A fixed 16-row tile read from an 8-row dynamic-M tensor is enough. No model config is involved; the script needs only `golden` on `PYTHONPATH`. The two kernels differ in one line, the row slice:

```python
x_tile = x[t0 : t0 + M_TILE, 0:K]                                    # faults
x_tile = pl.slice(x, [M_TILE, K], [t0, 0], valid_shape=[t_dim, K])   # passes
```

`K_ROW` is deliberately 65536 wide so the input tensor is large enough to be mmap-backed and the overread lands on a guard page every time. With a narrow row the same out-of-bounds read still happens but usually stays inside already-mapped memory and does not fault, so a passing run with small shapes does not mean the read is in bounds.

```python
"""min_hc_head_oob.py — fixed M-tile read past the end of a dynamic-M tensor.

python min_hc_head_oob.py -p a2a3sim         # Segmentation fault (rc=139)
python min_hc_head_oob.py -p a2a3sim --fix   # PASS
"""

import pypto.language as pl

T_DYN = pl.dynamic("T_DYN")

T = 8          # tokens actually present, mirrors DECODE_BATCH * DECODE_SEQ = 4 * 2
M_TILE = 16    # Cube fractal M, mirrors LINEAR_T_TILE
K_ROW = 65536  # full row width; large enough that x is mmap-backed
K = 256        # columns read per tile, mirrors LINEAR_K_TILE
N = 16         # mirrors HC_PAD


@pl.jit
def oob_linear_bug(
    x: pl.Tensor[[T_DYN, K_ROW], pl.FP32],
    w: pl.Tensor[[N, K], pl.FP32],
    out: pl.Out[pl.Tensor[[M_TILE, N], pl.FP32]],
):
    t_dim = pl.tensor.dim(x, 0)
    # The defect: the row count is rounded UP to the tile, and the slice below
    # then reads a whole M_TILE-row tile with no bound on the real row count.
    m_rows = pl.max(t_dim, M_TILE)
    with pl.spmd(m_rows // M_TILE, name_hint="oob_linear"):
        blk = pl.tile.get_block_idx()
        t0 = blk * M_TILE
        x_tile = x[t0 : t0 + M_TILE, 0:K]          # reads rows 0..15 of an 8-row tensor
        acc = pl.matmul(x_tile, w, b_trans=True, out_dtype=pl.FP32)
        out[0:M_TILE, 0:N] = acc
    return out


@pl.jit
def oob_linear_fixed(
    x: pl.Tensor[[T_DYN, K_ROW], pl.FP32],
    w: pl.Tensor[[N, K], pl.FP32],
    out: pl.Out[pl.Tensor[[M_TILE, N], pl.FP32]],
):
    t_dim = pl.tensor.dim(x, 0)
    m_rows = pl.max(t_dim, M_TILE)
    with pl.spmd(m_rows // M_TILE, name_hint="oob_linear"):
        blk = pl.tile.get_block_idx()
        t0 = blk * M_TILE
        x_tile = pl.slice(x, [M_TILE, K], [t0, 0], valid_shape=[t_dim, K])
        acc = pl.matmul(x_tile, w, b_trans=True, out_dtype=pl.FP32)
        out[0:M_TILE, 0:N] = acc
    return out


def golden(tensors):
    import torch

    x = tensors["x"].float()
    w = tensors["w"].float()
    ref = torch.zeros(M_TILE, N, dtype=torch.float32)
    ref[:T, :] = x[:, :K] @ w.T  # only the real rows are defined
    tensors["out"][:] = ref


def build_specs():
    import torch
    from golden import TensorSpec

    return [
        TensorSpec("x", [T, K_ROW], torch.float32, init_value=lambda: torch.randn(T, K_ROW) * 0.05),
        TensorSpec("w", [N, K], torch.float32, init_value=lambda: torch.randn(N, K) * 0.05),
        TensorSpec("out", [M_TILE, N], torch.float32, is_output=True),
    ]


if __name__ == "__main__":
    import argparse
    from golden import ratio_allclose, run_jit

    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", type=str, default="a2a3sim",
                        choices=["a2a3", "a2a3sim", "a5", "a5sim"])
    parser.add_argument("-d", "--device", type=int, default=0)
    parser.add_argument("--fix", action="store_true", help="use valid_shape on the row slice")
    args = parser.parse_args()

    result = run_jit(
        fn=oob_linear_fixed if args.fix else oob_linear_bug,
        specs=build_specs(),
        golden_fn=golden,
        runtime_cfg=dict(platform=args.platform, device_id=args.device),
        compare_fn={"out": ratio_allclose(atol=1e-3, rtol=1e-3, valid_rows=T)},
    )
    if not result.passed:
        if result.error:
            print(result.error)
        raise SystemExit(1)
```

### Reproduction of the reported case

```bash
python models/deepseek_v4_flash_mtp/hc_head.py -p a2a3sim   # Segmentation fault
python models/deepseek_v4_flash_mtp/hc_head.py -p a5sim     # Segmentation fault
```

Two one-line knobs in `models/deepseek_v4_flash_mtp/config.py` isolate the root cause without touching the kernel:

| `DECODE_BATCH` | `t_dim` | Relation to `LINEAR_T_TILE = 16` | Result on `a2a3sim` |
|---:|---:|---|---|
| 4 (default) | 8 | below the tile | Segmentation fault |
| 8 | 16 | exact multiple | PASS |
| 12 | 24 | above the tile, not a multiple | FAIL, `y` mismatch `26591/98304 = 27.05%` |

The third row is the important one: it is a deterministic numerical failure with no memory fault, which shows the defect is not merely an allocation-layout accident.

### Expected Behavior

- `hc_head` produces correct results for any token count, and never reads outside `x_flat`.
- A token count below `LINEAR_T_TILE` is handled as a partial tile, not by rounding the read up.
- A token count that is not a multiple of `LINEAR_T_TILE` still has every row computed.

### Actual Behavior

- `t_dim < LINEAR_T_TILE`: the Cube path reads about 449 KB past the end of `x_flat`, which segfaults on `a2a3sim` and `a5sim` and is silently tolerated on real A2/A3.
- `t_dim > LINEAR_T_TILE` and not a multiple of it: the trailing rows are never computed, `mixes_raw` keeps uninitialised values there, and the output is wrong by roughly 27 percent of elements at `t_dim = 24`.

### Fix

Apply the DSpark full-block / tail-block structure to the MTP linear stage.

```python
# row-block bounds
linear_full_rows = (t_dim // LINEAR_T_TILE) * LINEAR_T_TILE
t_linear = ((t_dim + LINEAR_T_TILE - 1) // LINEAR_T_TILE) * LINEAR_T_TILE

# aligned blocks keep the original static-M Cube path
if linear_full_rows > 0:
    with pl.spmd((linear_full_rows // LINEAR_T_TILE) * LINEAR_OK, name_hint="hc_head_linear", ...):
        ...
        x_lin_full = x_flat[t0 : t0 + LINEAR_T_TILE, k0 : k0 + LINEAR_K_TILE]

# at most one incomplete block, read with an explicit valid_shape
if linear_full_rows < t_dim:
    with pl.spmd(LINEAR_OK, name_hint="hc_head_linear_tail", ...):
        ...
        tail_rows = t_dim - linear_full_rows
        x_lin_tail = pl.slice(x_flat, [LINEAR_T_TILE, LINEAR_K_TILE],
                              [linear_full_rows, k0],
                              valid_shape=[tail_rows, LINEAR_K_TILE])
```

This mirrors `models/deepseek_v4_flash_dspark/hc_head.py` exactly, so the two variants stay consistent. An alternative, lighter change is to add `valid_shape` to the existing `x_lin` slice, but that alone does not fix the dropped-rows case above, because `t_linear` would still not be a multiple of the tile.

A complementary hardening option, outside this issue: have pypto emit a `valid_shape` (or clamp) automatically when a slice extends past a tensor's real extent, so this class of defect fails loudly at compile time instead of depending on memory layout.

### Test Results

Fix applied to `models/deepseek_v4_flash_mtp/hc_head.py` only, 68 insertions and 19 deletions. Verified with pto-isa carrying the current simulator fixes (gitcode `cann/pto-isa` MRs 1580, 1581, 1583).

| Case | Platform | Before | After |
|---|---|---|---|
| `models/deepseek_v4_flash_mtp/hc_head.py` (`t_dim = 8`) | `a2a3sim` | Segmentation fault, rc=139 | **PASS 4.83 s**, `y` validated |
| `models/deepseek_v4_flash_mtp/hc_head.py` (`t_dim = 8`) | `a5sim` | Segmentation fault, rc=139 | **PASS 4.98 s**, `y` validated |
| `models/deepseek_v4_flash_mtp/hc_head.py` (`t_dim = 24`, `DECODE_BATCH=12`) | `a2a3sim` | FAIL, `y` `26591/98304 = 27.05%` over tolerance | **PASS 4.96 s** |
| `models/deepseek_v4_flash_mtp/decode_mtp.py` (only importer in the sim sweep) | `a2a3sim` | `async-DMA workspace provisioning (mask=1) failed with code -1001` | compile 4.27 s OK, same pre-existing `-1001`, no new regression |
| `models/deepseek_v4_flash_dspark/hc_head.py` (untouched control) | `a2a3sim` | PASS | PASS 5.79 s |
| minimal reproduction, no `valid_shape` | `a2a3sim`, `a5sim` | Segmentation fault, rc=139 | n/a |
| minimal reproduction, with `valid_shape` | `a2a3sim`, `a5sim` | n/a | **PASS** |
| `ruff check models/deepseek_v4_flash_mtp/hc_head.py` | — | — | All checks passed |

The `DECODE_BATCH` edits used for the `t_dim = 24` experiment were reverted; only `hc_head.py` is modified.

### Environment

| Component | Version |
|---|---|
| pypto-lib CI commit | `0e7e07aa5e42db6e08c68fd5bfe1def571492b53` |
| pypto CI commit | `cb49d101aa770b72c0113f6a9ee32301f38464ff` |
| simpler CI commit (`pypto/runtime`) | `dbdd041e957420ea15b03e878400dd4de5e9c34c` |
| pto-isa | `master` + gitcode `cann/pto-isa` MRs 1580 / 1581 / 1583 |
| PTOAS | `v0.57` (`aarch64`) |
| Simulator targets | `a2a3sim`, `a5sim` |
| Host platform | Linux (`aarch64`) |

### Additional Context

- This was originally filed against pto-isa as a simulator segfault of unknown origin. pto-isa cannot fix it: the generated `GlobalTensor` declares a 16-row region and the real tensor extent is never passed down, so the CPU backend has no way to know the read is out of bounds. The tracking note has been moved to pypto-lib.
- The same tile-rounding pattern is worth auditing elsewhere in the MTP model set. `pl.max(t_dim, TILE)` guarantees only "not smaller than the tile", never "a multiple of the tile", so any use of it as a loop bound over `TILE`-row blocks has the same two failure modes.
- The DSpark reduce stage additionally clamps per-tile rows with `valid_rows = pl.min(T_TILE, t_dim - t0)`; the MTP reduce stage assumes `t_dim` is a multiple of `T_TILE = 8`. That holds for the current decode configuration and is out of scope here, but it is the next thing to check if MTP is ever run with a token count that is not a multiple of 8.
